components/supervisor/pkg/gitpod/gitpod-config.go

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package gitpod
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ghodss/yaml"
+	"github.com/gitpod-io/gitpod/common-go/log"
+)
+
+// ConfigInterface provides an acess to the gitpod config file.
+type ConfigInterface interface {
+	// Observe provides channels triggered whenever the config is changed or errored
+	Observe(ctx context.Context) (<-chan *GitpodConfig, <-chan error)
+}
+
+// ConfigService provides an acess to the gitpod config file.
+type ConfigService struct {
+	location      string
+	locationReady <-chan struct{}
+
+	config    *GitpodConfig
+	listeners map[configListener]struct{}
+	stop      context.CancelFunc
+	mutex     sync.Mutex
+	timer     *time.Timer
+}
+
+type configListener struct {
+	configs chan *GitpodConfig
+	errors  chan error
+}
+
+// NewConfigService creates a new instance of ConfigService
+func NewConfigService(configLocation string, locationReady <-chan struct{}) *ConfigService {
+	return &ConfigService{
+		location:      configLocation,
+		locationReady: locationReady,
+		listeners:     make(map[configListener]struct{}),
+	}
+}
+
+// Observe provides channels triggered whenever the config is changed or errored
+func (service *ConfigService) Observe(ctx context.Context) (<-chan *GitpodConfig, <-chan error) {
+	listener := configListener{
+		configs: make(chan *GitpodConfig),
+		errors:  make(chan error),
+	}
+
+	go func() {
+		defer close(listener.configs)
+		defer close(listener.errors)
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-service.locationReady:
+		}
+
+		err := service.start()
+		if err != nil {
+			// failed to start
+			listener.errors <- err
+			return
+		}
+		listener.configs <- service.config
+
+		service.mutex.Lock()
+		service.listeners[listener] = struct{}{}
+		service.mutex.Unlock()
+
+		defer func() {
+			service.mutex.Lock()
+			defer service.mutex.Unlock()
+			delete(service.listeners, listener)
+			if len(service.listeners) == 0 && service.stop != nil {
+				service.stop()
+				service.stop = nil
+			}
+		}()
+
+		select {
+		case <-ctx.Done():
+			return
+		}
+	}()
+	return listener.configs, listener.errors
+}
+
+func (service *ConfigService) start() error {
+	context := service.tryStart()
+	if context == nil {
+		// alread running
+		return nil
+	}
+	_, err := os.Stat(service.location)
+	if service.tryPolling(context, err) {
+		// inotify cannot watch inexistent file, let's poll
+		return nil
+	}
+	return service.watch(context)
+}
+
+func (service *ConfigService) tryStart() context.Context {
```

Comment:
It's a bit unusual that a function returns a `context.Context`.
Also, it makes `start()` harder to understand because one needs to mentally bring `tryStart` and `start` together.
Why not unite them into one function?